### PR TITLE
[Test]: Add INT8 (W8A8) quantization support and A/B gates for MammothModa2

### LINKED ISSUE #7125: [Performance]: Enable supported weight/activation quantization formats and validate generation experts, extra vocabulary/head, checkpoint loading, and quantization scales.
### Proposal to improve performance

Parent: #7075 — sub-issue for the AR-stage quantization TODO.

Enable FP8 (dynamic W8A8) and INT8 (W8A8) quantization for the MammothModa2 AR stage to reduce AR prefill/decode latency and peak GPU memory.

## Owner and scope

**Owner:** @wenj-yan

**In scope (MammothModa2 AR stage):**

- Enable supported weight/activation quantization formats for the AR LLM, starting with FP8 (dynamic W8A8) and INT8 (W8A8), via the unified `build_quant_config()` entry (#1763 / #1764).
- Validate **generation experts** (generation-specific expert/FFN paths are quantized correctly; routers/gates stay in full precision per the convention in #1854).
- Validate the **extra vocabulary / lm_head**: correct quantization exclusion and prefix handling for the extended vocab and generation head, so exclude-lists actually hit (precedent: lm_head prefix fix in #4528; unquantized lm_head GEMM issue in #1790).
- Validate **checkpoint loading** for pre-quantized checkpoints: weight/scale loading, packed-module mapping, and scale routing to the AR component only (precedent for wrong-scope routing: #2686; scale remapping: #5087).
- Validate **quantization scales**: scale dtype/shape sanity, NaN/clamp handling on load.
- Document the supported AR quantization configurations in the per-stage support matrix.

**Exclusions (owned elsewhere or out of scope):**

- DiT/transformer quantization — @xRay2016 (#7075 DiT quantization TODO).
- Quantized AR KV cache — @JiahengX.

## Plan

- **PR plan:**
  - PR 1 — Enable online FP8 (W8A8 dynamic) for the AR stage: quant_config plumbing, generation-expert handling, lm_head/extra-vocab exclusion, unit tests.
  - PR 2 — INT8 (W8A8) + pre-quantized checkpoint loading: scale loading/remapping, scale validation, loader tests.
  - PR 3 — Validation & docs: quality/parity test suite, support-matrix update.
- **PR ready for review:** 2026-09-14 (UTC)
- **Target merge:** 2026-09-17 (UTC), in line with the umbrella checkpoint.

## Correctness / quality validation

- Numerical-parity tests for quantized (FP8/INT8) vs BF16 AR logits on mixed-token routing and generation constraints.
- AR-stage output (logits / hidden states) alignment checks.
- Checkpoint-loading tests: pre-quantized INT8 checkpoints load without scale errors; quantized weights land only on AR modules.

## Known blockers / constraints

- Pre-quantized INT8/FP8 MammothModa2 checkpoints may need to be produced locally (no official quantized checkpoint known).
- MammothModa2 has two versions, Preview and Dev, with different AR-LLM architectures. Both must be validated independently as separate baselines.

## Reviewer

Requesting maintainer review.

Refs #7075

### Report of performance regression

N/A

## Hardware / baseline / performance validation plan

- **Hardware/backend:** NVIDIA CUDA, RTX PRO 6000 / 96 GB ×1.
- **Model version note:** MammothModa2 has two versions, Preview and Dev, with different AR-LLM architectures. Both must be validated independently as separate baselines.
- **Baseline:**
  - BF16 baseline from @quan-yi-ai, using AR logits/hidden-state outputs as reference.
  - Preview and Dev versions baselined separately; AR quantization validation covers both.
- **Performance evidence (AR-only scope):**
  - AR prefill/decode latency and TPS, peak GPU memory.
  - A/B comparison against the BF16 baseline from @quan-yi-ai.
  - Preview and Dev versions recorded separately.

### Your current environment (if you think it is necessary)

```text
The output of `python collect_env.py`
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by wenj-yan**
Resolves #7125 (INT8 portion). Part of #7075. Follow-up to #7334 (PR 1, FP8).

## Purpose

Enable INT8 (W8A8) quantization for the MammothModa2 AR stage, with pre-quantized checkpoint loading support (weight/scale loading, packed-module mapping, and scale routing to the AR stage only), validated against the published MammothModa2-Preview/Dev-W8A8 checkpoints (see Quantized Checkpoints below). This corresponds to PR 2 of the plan in #7125.
Tests added in `test_mammoth_moda2_int8_quantization.py` (GPU E2E A/B gates,
BF16 vs INT8 under the same seed with greedy decoding):

- `test_ar_generation_smoke[bf16|int8]` — each format loads and produces a  non-empty greedy sequence;
- `test_bf16_vs_int8_generation_consistency` — AR text A/B: token agreement,  agree-prefix length, top-1 logprob cosine / MAE;
- `test_bf16_vs_int8_t2i_image_consistency` — t2i image A/B with INT8 applied  to the AR stage only (stage 0), DiT kept in BF16: decoded-image cosine,  rel-L2, PSNR, MAE, max-abs, pixel-match.

Per #7125, both MammothModa2 versions (Preview and Dev, which have different AR-LLM architectures) are validated independently.

## Quantized Checkpoints

Pre-quantized INT8 (W8A8) weights used by the tests are published on ModelScope:

- Preview: [MammothModa2-Preview-W8A8](https://modelscope.cn/models/<org>/MammothModa2-Preview-W8A8)
- Dev: [MammothModa2-Dev-W8A8](https://modelscope.cn/models/<org>/MammothModa2-Dev-W8A8)

## Test Plan

```
pytest tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py -v -s
```

All cases use seed=42 and greedy decoding; the only A/B variable is the
stage-0 quantization key.

**vLLM Version:** 0.28.0
**Hardware:** NVIDIA RTX 4090 48 GB ×1, Python 3.12

### MammothModa2-Preview

| Test | Metric | Measured |
| --- | --- | --- |
| AR text A/B | token agreement | 0.2188 |
| | agree prefix | 7/32 |
| | top-1 logprob cosine | 0.9996 |
| | logprob MAE | 0.0592 |
| t2i image A/B | PSNR | 40.13 dB |
| | image cosine | 0.999796 |
| | rel-L2 | 0.022230 |
| | MAE / max-abs / pixel-match | 0.0073 / 0.0745 / 52.5% |

### MammothModa2-Dev

| Test | Metric | Measured |
| --- | --- | --- |
| t2i image A/B | PSNR | 33.50 dB |
| | image cosine | 0.999847 |
| | rel-L2 | 0.038357 |
| | MAE / max-abs / pixel-match | 0.0195 / 0.0902 / 5.4% |

### Conclusion

- On both **Preview** and **Dev**, decoded-image quality is preserved:  image cosine ≥ 0.9998 and rel-L2 < 0.05 on both versions.
- **Preview** reaches 40.1 dB PSNR with 52.5% exact pixel match; top-1  logprob cosine 0.9996. Greedy token agreement is low (0.2188, common  prefix 7/32) — INT8 logit perturbations flip low-margin tokens, but the  logprob distribution stays well aligned.
- **Dev** image A/B is coarser (PSNR 33.5 dB, pixel-match 5.4%) while  remaining within the cosine / rel-L2 gates; per-pixel differences are  larger than on Preview.

### Notes

- Unlike FP8 (#7334), greedy token sequences diverge under INT8 on Preview  (agreement 0.2188 vs 1.0 for FP8); gates for INT8 are therefore set on  logprob-level and image-level metrics rather than exact token match.
- Note: vLLM's INT8 W8A8 backend does not support SM100/Blackwell  (compute capability ≥ 10.0, e.g. RTX PRO 6000); tests were therefore run on  RTX 4090 48 GB (Ada, CC 8.9), and Blackwell users should fall back to  FP8 (#7334) or NVFP4.
- DiT quantization is owned by @xRay2016 under #7075.
- Validation-suite hardening and the per-stage support-matrix update are  left to PR 3 per the plan in #7125.

### CODE REVIEW

**Comment by lishunyang12** [comment]
is quantization already supported?

**Comment by wenj-yan** [comment]
Yeah, there was already a path to pass the quantization_config. It inherits the quantization from qwen2.5 and qwen3.

File: tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+    logprob_mae = _mean_abs_diff(bf16_lp[:lp_len], int8_lp[:lp_len]) if lp_len > 0 else float("nan")
+
+    print(
+        f"[INT8 A/B] token_agreement={token_agree:.4f} "
+        f"agree_prefix={agree_prefix}/{common} "
+        f"logprob_cosine={logprob_cos:.4f} "
+        f"logprob_mae={logprob_mae:.4f}"
+    )
+
+    if logprob_cos == logprob_cos:  # not NaN
```
Comment (lishunyang12) [inline tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py]: Fail when the shared prefix is too short; NaN must not bypass the quality gate.

File: tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+# Constants
+# ---------------------------------------------------------------------------
+
+# BF16 baseline checkpoint (same one the FP8 A/B gate uses).
+MODEL_PATH = os.environ.get("MAMMOTH_MODA2_MODEL", "bytedance-research/MammothModa2-Dev")
+
+# Serialized INT8 W8A8 checkpoint (compressed-tensors). Override via env for CI
+# or hub ids; the default points at the locally produced llm-compressor
+# checkpoint.
+INT8_MODEL_PATH = os.environ.get("MAMMOTH_MODA2_INT8_MODEL", "/root/autodl-fs/MammothModa2-Dev-W8A8")
```
Comment (lishunyang12) [inline tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py]: Use a published checkpoint ID instead of this workstation-specific path.

File: tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+    for candidate in (model, MODEL_PATH):
+        local = Path(candidate) / "t2i_generation_config.json"
+        if local.exists():
+            return json.loads(local.read_text(encoding="utf-8"))
+
+    from huggingface_hub import snapshot_download
+
+    for candidate in (model, MODEL_PATH):
+        try:
+            weights_dir = Path(snapshot_download(candidate))
```
Comment (lishunyang12) [inline tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py]: Download only t2i_generation_config.json instead of entire checkpoints.

File: tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py (hunk 2 above)
Comment (wenj-yan) [reply tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py]: The official checkpoint doesn't contain a quantized model, so I put my own quantized model up on Hugging Face and ModelScope.Fix in 30651e6

File: tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py (hunk 1 above)
Comment (wenj-yan) [reply tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py]: Replaced the NaN-skipped logprob check with a hard`MIN_SHARED_PREFIX = 2` assert plus an unconditional cosine assert.fix in 30651e6

File: tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py (hunk 3 above)
Comment (wenj-yan) [reply tests/e2e/offline_inference/test_mammoth_moda2_int8_quantization.py]: Thank you for your review!I have switched the t2i config lookup from`snapshot_download` to a single-file`hf_hub_download` (via`repo_utils.hf_api` ).Fix in 30651e6.
For the new modifications, I re-ran the tests, and the results are as follows:
| Category | Metric | Value |
|---|---|---|
| **[INT8 A/B]** | token_agreement | 0.9375 |
|  | agree_prefix | 12/32 |
|  | logprob_cosine | 0.9768 |
|  | logprob_mae | 0.0373 |
| **[INT8 t2i A/B]**<br>BF16 vs INT8 decoded image | psnr_db | 33.496779 |
|  | mae | 0.019488 |
|  | max_abs | 0.090196 |
|  | cosine | 0.999847 |
|  | rel_l2 | 0.038357 |
|  | pixel_match | 0.053772 |
